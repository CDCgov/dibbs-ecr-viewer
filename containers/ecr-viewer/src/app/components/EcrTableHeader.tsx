"use client";
import React from "react";
import { useQueryParam } from "../hooks/useQueryParam";
import { SortButton } from "./SortButton";

export type TableHeader = {
  id: string;
  value: string;
  className: string;
  dataSortable: boolean;
  sortDirection: string;
};

/**
 * Interactive header for the ecr library table
 * @param params react params
 * @param params.headers header descriptions
 * @param params.disabled whether to disable the sort functionality
 * @returns Interactive header row
 */
export const EcrTableHeader = ({
  headers,
  disabled,
}: {
  headers: TableHeader[];
  disabled: boolean;
}) => {
  const { updateQueryParam, pushQueryUpdate } = useQueryParam();

  /**
   * Handles sorting the table data by a given column. We update the search params,
   * which triggers a re-render of this component with the updated props when the
   * page gets the new search params.
   * @param columnId - The ID of the column to sort by.
   * @param curDirection - The current direction of sort.
   */
  const handleSort = (columnId: string, curDirection: string) => {
    // Flip the sort from the current direction, ASC is default
    const direction = curDirection === "ASC" ? "DESC" : "ASC";

    updateQueryParam("columnId", columnId);
    updateQueryParam("direction", direction);
    pushQueryUpdate();
  };

  return (
    <thead>
      <tr>
        {headers.map((column) => (
          <Header
            key={column.id}
            column={column}
            handleSort={handleSort}
            disabled={disabled}
          />
        ))}
      </tr>
    </thead>
  );
};

type SortHandlerFn = (columnId: string, direction: string) => void;

const Header = ({
  column,
  disabled,
  handleSort,
}: {
  column: TableHeader;
  disabled: boolean;
  handleSort: SortHandlerFn;
}) => {
  return (
    <th
      id={`${column.id}-header`}
      scope="col"
      role="columnheader"
      className={column.className}
      data-sortable={column.dataSortable}
      aria-sort={getAriaSortValue(column.sortDirection)}
    >
      {column.sortDirection || column.dataSortable ? (
        <SortButton
          columnId={column.id}
          columnName={column.value}
          direction={column.sortDirection}
          disabled={disabled}
          handleSort={() => handleSort(column.id, column.sortDirection)}
        />
      ) : (
        <div className="display-flex">{column.value}</div>
      )}
    </th>
  );
};

type AriaSortType = "none" | "ascending" | "descending" | "other";

const getAriaSortValue = (sortDirection: string): AriaSortType | undefined => {
  if (sortDirection === "ASC") {
    return "ascending";
  } else if (sortDirection === "DESC") {
    return "descending";
  }
};
