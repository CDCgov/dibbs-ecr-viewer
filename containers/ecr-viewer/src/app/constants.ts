export const INITIAL_HEADERS = [
  {
    id: "patient",
    value: "Patient",
    className: "library-patient-column",
    dataSortable: true,
    sortDirection: "",
  },
  {
    id: "date_created",
    value: "Received Date",
    className: "library-received-date-column",
    dataSortable: true,
    sortDirection: "",
  },
  {
    id: "encounter_date",
    value: "Encounter Date",
    className: "library-encounter-date-column",
    dataSortable: true,
    sortDirection: "",
  },
  {
    id: "reportable_condition",
    value: "Reportable Condition",
    className: "library-condition-column",
    dataSortable: false,
    sortDirection: "",
  },
  {
    id: "rule_summary",
    value: "RCKMS Rule Summary",
    className: "library-rule-column",
    dataSortable: false,
    sortDirection: "",
  },
];
