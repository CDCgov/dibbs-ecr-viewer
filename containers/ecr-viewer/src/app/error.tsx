"use client"; // Error components must be Client Components

import { useEffect } from "react";
import ErrorPage from "@/app/components/ErrorPage";

/**
 * Renders an Error component to display when an error occurs.
 * @param props - The props object.
 * @param props.error - The error object, possibly with a digest.
 * @returns - Returns the JSX element for the Error component.
 */
export default function Error({
  error,
}: {
  error: Error & { digest?: string };
}) {
  useEffect(() => {
    // Log the error to an error reporting service
    console.error(error);
  }, [error]);

  return (
    <ErrorPage title="Something went wrong!">
      If the problem persists, please reach out to your eCR coordinator.
    </ErrorPage>
  );
}
