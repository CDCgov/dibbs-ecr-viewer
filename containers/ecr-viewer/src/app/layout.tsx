import "@/styles/styles.scss";
import { PublicEnvScript, env } from "next-runtime-env";

export const metadata = {
  title: "DIBBs eCR Viewer",
  description: "View your eCR data in an easy-to-understand format.",
};

const PATIENT_BANNER_BUFFER = "2.75rem";

declare module "react" {
  interface CSSProperties {
    "--patient-banner-buffer"?: typeof PATIENT_BANNER_BUFFER | 0;
  }
}

/**
 * `RootLayout` serves as the top-level layout component for a React application.
 * @param props - The properties passed to the component.
 * @param props.children - The child components or elements to be rendered within
 *   the `<body>`f tag of the HTML document.
 * @returns A React element representing the top-level HTML structure, with the
 *   `children` rendered inside the `<body>` tag.
 */
export default function RootLayout({
  children,
}: {
  children: React.ReactNode;
}) {
  const isNonIntegratedViewer =
    env("NEXT_PUBLIC_NON_INTEGRATED_VIEWER") === "true";

  return (
    <html
      lang="en"
      style={{
        "--patient-banner-buffer": isNonIntegratedViewer
          ? PATIENT_BANNER_BUFFER
          : 0,
      }}
    >
      <head>
        <PublicEnvScript nonce={{ headerKey: "x-nonce" }} />
      </head>
      <body className="overflow-x-auto">{children}</body>
    </html>
  );
}
