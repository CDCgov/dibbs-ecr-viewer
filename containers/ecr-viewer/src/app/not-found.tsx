import ErrorPage from "@/app/components/ErrorPage";
import { env } from "next-runtime-env";

/**
 * 404 page
 * @returns 404 Page
 */
const NotFound = () => {
  const isNonIntegratedViewer =
    env("NEXT_PUBLIC_NON_INTEGRATED_VIEWER") === "true";

  return (
    <ErrorPage
      title="Page not found"
      subTitle="The requested page could not be found"
    >
      Please try the following:
      <ul className="margin-0 padding-left-3">
        <li>
          <b>Check the URL:</b> Make sure there are no typos in the address.
        </li>
        {!isNonIntegratedViewer && (
          <li>
            <b>Return to NBS:</b> Return to NBS and try to reopen the eCR.
          </li>
        )}
        <li>
          <b>Contact support:</b> If the problem persists, please reach out to
          your eCR coordinator.
        </li>
      </ul>
    </ErrorPage>
  );
};

export default NotFound;
