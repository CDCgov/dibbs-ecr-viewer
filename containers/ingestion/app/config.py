from functools import lru_cache
from typing import Literal, Optional

from pydantic_settings import BaseSettings


class Settings(BaseSettings):
    cred_manager: Optional[Literal["azure", "gcp"]] = None
    salt_str: Optional[str] = None
    fhir_url: Optional[str] = None
    smarty_auth_id: Optional[str] = None
    smarty_auth_token: Optional[str] = None
    license_type: Optional[str] = None
    cloud_provider: Optional[Literal["azure", "gcp"]] = None
    bucket_name: Optional[str] = None
    storage_account_url: Optional[str] = None


@lru_cache
def get_settings() -> dict:
    """
    Load the values specified in the Settings class from the environment and return a
    dictionary containing them. The dictionary is cached to reduce overhead accessing
    these values.

    :return: A dictionary with keys specified by the Settings. The value of each key is
    read from the corresponding environment variable.
    """
    return Settings().model_dump()
