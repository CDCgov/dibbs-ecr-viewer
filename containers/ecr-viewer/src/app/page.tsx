import React, { Suspense } from "react";
import Header from "@/app/components/Header";
import { getTotalEcrCount } from "@/app/services/listEcrDataService";
import EcrPaginationWrapper from "@/app/components/EcrPaginationWrapper";
import EcrTable from "@/app/components/EcrTable";
import LibrarySearch from "./components/LibrarySearch";
import Filters from "@/app/components/Filters";
import { EcrTableLoading } from "./components/EcrTableClient";
import { returnParamDates } from "@/app/utils/date-utils";
import { env } from "next-runtime-env";
import { getAllConditions } from "./data/conditions";
import NotFound from "@/app/not-found";

/**
 * Functional component for rendering the home page that lists all eCRs.
 * @param props - parameters from the HomePage
 * @param props.searchParams - list of search params
 * @returns The home page JSX component.
 */
const HomePage = async ({
  searchParams,
}: {
  searchParams: { [key: string]: string | string[] | undefined };
}) => {
  const isNonIntegratedViewer =
    env("NEXT_PUBLIC_NON_INTEGRATED_VIEWER") === "true";

  if (!isNonIntegratedViewer) {
    return <NotFound />;
  }

  const currentPage = Number(searchParams?.page) || 1;
  const itemsPerPage = Number(searchParams?.itemsPerPage) || 25;
  const sortColumn = (searchParams?.columnId as string) || "date_created";
  const sortDirection = (searchParams?.direction as string) || "DESC";
  const searchTerm = searchParams?.search as string | undefined;
  const filterConditions = searchParams?.condition as string | undefined;
  const filterConditionsArr = filterConditions?.split("|");
  const filterDates = returnParamDates(searchParams);

  const totalCount = await getTotalEcrCount(
    filterDates,
    searchTerm,
    filterConditionsArr,
  );

  const allConditions = await getAllConditions();

  return (
    <div className="display-flex flex-column height-viewport">
      <Header />
      <main className="overflow-auto height-full">
        <div className="margin-x-3 padding-y-105 display-flex flex-align-center">
          <h2 className="margin-bottom-0 text-bold font-sans-xl">
            eCR Library
          </h2>
          <LibrarySearch
            className="margin-left-auto"
            textBoxClassName="width-21-9375"
          />
        </div>
        <Filters conditions={allConditions} />
        <EcrPaginationWrapper totalCount={totalCount}>
          <Suspense fallback={<EcrTableLoading />}>
            <EcrTable
              currentPage={currentPage}
              itemsPerPage={itemsPerPage}
              sortColumn={sortColumn}
              sortDirection={sortDirection}
              searchTerm={searchTerm}
              filterConditions={filterConditionsArr}
              filterDates={filterDates}
            />
          </Suspense>
        </EcrPaginationWrapper>
      </main>
    </div>
  );
};

export default HomePage;
