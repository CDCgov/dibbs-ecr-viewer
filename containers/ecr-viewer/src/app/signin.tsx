"use client";

import { Button } from "@trussworks/react-uswds";
import { ArrowForward } from "@/app/components/Icon";
import bgRedirect from "../../assets/bg-redirect.png";
import { signIn } from "next-auth/react";
import { ProviderDetails } from "./api/auth/auth";

interface RedirectProps {
  provider: ProviderDetails;
}
/**
 * Redirect Component
 *
 * Returns a sign-in page for users trying to access the eCR Viewer.
 * @param props - The props object.
 * @param props.provider - Information about the provider.
 * @returns A styled div with prompt message and a login button that redirects users their authentication provider.
 */
export const Redirect = ({ provider }: RedirectProps) => {
  return (
    <div
      className="position-relative text-white"
      style={{
        backgroundImage: `url(${bgRedirect.src})`,
        backgroundSize: "cover",
        backgroundPosition: "center",
        minHeight: "100vh",
        minWidth: "100vw",
        overflowY: "auto",
        overflowX: "auto",
      }}
    >
      <div className="width-tablet position-absolute top-3950 left-12 padding-bottom-6 padding-right-6 grid-row gap-5">
        <h1 className="font-serif-3xl text-bold margin-0">
          Looks like you're trying to access the eCR Viewer
        </h1>
        <p className="font-sans-lg text-bold margin-0">
          You need to log in to see the eCR Viewer
        </p>
        <br />
        <Button
          className="redirect-button"
          type="button"
          onClick={async () => {
            await signIn(provider.id, { callbackUrl: process.env.BASE_PATH });
          }}
        >
          Log in via {provider.name}
          <ArrowForward aria-hidden={true} size={3} />
        </Button>
      </div>
    </div>
  );
};
