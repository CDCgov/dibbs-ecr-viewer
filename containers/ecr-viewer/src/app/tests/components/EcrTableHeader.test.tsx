import { axe } from "jest-axe";
import { act, render } from "@testing-library/react";
import EcrTableHeader from "@/app/components/EcrTableHeader";
import { INITIAL_HEADERS } from "@/app/constants";
import router from "next-router-mock";

jest.mock("next/navigation", () => ({
  useRouter: () => router,
  usePathname: () => router.pathname,
  useSearchParams: () => new URLSearchParams(router.asPath.split("?")[1] || ""),
}));

describe("EcrTableHeader", () => {
  beforeEach(() => {
    jest.resetAllMocks();
    router.setCurrentUrl(
      "/?itemsPerPage=25&columnId=report_date&direction=ASC&page=1",
    );
  });

  describe("load with an eCR", () => {
    it("should match snapshot", async () => {
      const table = document.createElement("table");
      const { container } = render(
        <EcrTableHeader headers={INITIAL_HEADERS} disabled={false} />,
        {
          container: document.body.appendChild(table),
        },
      );
      expect(container).toMatchSnapshot();
    });

    it("should pass accessibility", async () => {
      const table = document.createElement("table");
      const { container } = render(
        <EcrTableHeader headers={INITIAL_HEADERS} disabled={false} />,
        {
          container: document.body.appendChild(table),
        },
      );
      await act(async () => {
        expect(await axe(container)).toHaveNoViolations();
      });
    });
  });
});
