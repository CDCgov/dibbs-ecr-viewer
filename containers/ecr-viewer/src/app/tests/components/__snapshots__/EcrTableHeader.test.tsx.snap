// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EcrTableHeader disabled should match snapshot 1`] = `
<table>
  <thead>
    <tr>
      <th
        class="library-patient-column"
        data-sortable="true"
        id="patient-header"
        role="columnheader"
        scope="col"
      >
        <button
          aria-label="Sort by Patient"
          class="usa-button usa-button--unstyled sort-button usa-button"
          data-testid="button"
          disabled=""
          id="patient-sort-button"
          type="button"
        >
          <span
            class="text-ink"
          >
            Patient
          </span>
          <svg
            aria-hidden="true"
            class="usa-icon square-3"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.17 15 13 17.17V6.83L15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 11 6.83v10.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15z"
            />
          </svg>
        </button>
      </th>
      <th
        class="library-received-date-column"
        data-sortable="true"
        id="date_created-header"
        role="columnheader"
        scope="col"
      >
        <button
          aria-label="Sort by Received Date"
          class="usa-button usa-button--unstyled sort-button usa-button"
          data-testid="button"
          disabled=""
          id="date_created-sort-button"
          type="button"
        >
          <span
            class="text-ink"
          >
            Received Date
          </span>
          <svg
            aria-hidden="true"
            class="usa-icon square-3"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.17 15 13 17.17V6.83L15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 11 6.83v10.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15z"
            />
          </svg>
        </button>
      </th>
      <th
        class="library-encounter-date-column"
        data-sortable="true"
        id="encounter_date-header"
        role="columnheader"
        scope="col"
      >
        <button
          aria-label="Sort by Encounter Date"
          class="usa-button usa-button--unstyled sort-button usa-button"
          data-testid="button"
          disabled=""
          id="encounter_date-sort-button"
          type="button"
        >
          <span
            class="text-ink"
          >
            Encounter Date
          </span>
          <svg
            aria-hidden="true"
            class="usa-icon square-3"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.17 15 13 17.17V6.83L15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 11 6.83v10.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15z"
            />
          </svg>
        </button>
      </th>
      <th
        class="library-condition-column"
        data-sortable="false"
        id="reportable_condition-header"
        role="columnheader"
        scope="col"
      >
        <div
          class="display-flex"
        >
          Reportable Condition
        </div>
      </th>
      <th
        class="library-rule-column"
        data-sortable="false"
        id="rule_summary-header"
        role="columnheader"
        scope="col"
      >
        <div
          class="display-flex"
        >
          RCKMS Rule Summary
        </div>
      </th>
    </tr>
  </thead>
</table>
`;

exports[`EcrTableHeader enabled should match snapshot 1`] = `
<table>
  <thead>
    <tr>
      <th
        class="library-patient-column"
        data-sortable="true"
        id="patient-header"
        role="columnheader"
        scope="col"
      >
        <button
          aria-label="Sort by Patient"
          class="usa-button usa-button--unstyled sort-button usa-button"
          data-testid="button"
          id="patient-sort-button"
          type="button"
        >
          <span
            class="text-ink"
          >
            Patient
          </span>
          <svg
            aria-hidden="true"
            class="usa-icon square-3"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.17 15 13 17.17V6.83L15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 11 6.83v10.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15z"
            />
          </svg>
        </button>
      </th>
      <th
        class="library-received-date-column"
        data-sortable="true"
        id="date_created-header"
        role="columnheader"
        scope="col"
      >
        <button
          aria-label="Sort by Received Date"
          class="usa-button usa-button--unstyled sort-button usa-button"
          data-testid="button"
          id="date_created-sort-button"
          type="button"
        >
          <span
            class="text-ink"
          >
            Received Date
          </span>
          <svg
            aria-hidden="true"
            class="usa-icon square-3"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.17 15 13 17.17V6.83L15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 11 6.83v10.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15z"
            />
          </svg>
        </button>
      </th>
      <th
        class="library-encounter-date-column"
        data-sortable="true"
        id="encounter_date-header"
        role="columnheader"
        scope="col"
      >
        <button
          aria-label="Sort by Encounter Date"
          class="usa-button usa-button--unstyled sort-button usa-button"
          data-testid="button"
          id="encounter_date-sort-button"
          type="button"
        >
          <span
            class="text-ink"
          >
            Encounter Date
          </span>
          <svg
            aria-hidden="true"
            class="usa-icon square-3"
            focusable="false"
            height="1em"
            role="img"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M15.17 15 13 17.17V6.83L15.17 9l1.41-1.41L12 3 7.41 7.59 8.83 9 11 6.83v10.34L8.83 15l-1.41 1.41L12 21l4.59-4.59L15.17 15z"
            />
          </svg>
        </button>
      </th>
      <th
        class="library-condition-column"
        data-sortable="false"
        id="reportable_condition-header"
        role="columnheader"
        scope="col"
      >
        <div
          class="display-flex"
        >
          Reportable Condition
        </div>
      </th>
      <th
        class="library-rule-column"
        data-sortable="false"
        id="rule_summary-header"
        role="columnheader"
        scope="col"
      >
        <div
          class="display-flex"
        >
          RCKMS Rule Summary
        </div>
      </th>
    </tr>
  </thead>
</table>
`;
